rust-symcrypt/src/hash.rs

```diff
@@ -1,59 +1,89 @@
+/// 
+/// result() returns the result of the hash. Once result() is called, the lifetime of the ShaXXXState is finished.
+/// To perform other stateful hash operation you must create a new hash object via ShaXXXState::new()
+/// 
+/// copy() creates a copy of the current ShaXXXState
+pub trait HashState {
     type Result;
 
     fn append(&mut self, data: &[u8]);
-    fn result(&mut self) -> Self::Result;
+    fn result(self) -> Self::Result;
+    fn copy(&self) -> Box<Self>;
 }
 
+/// Sha256State needs to have a heap allocated inner state that is Pin<Box<T>> since the memory address of Self is moved around when implementing 
+/// HashState Result field.
+/// 
+/// SymCrypt expects the address for its structs to stay static through the structs lifetime to guarantee that structs are not memcpy'd as
+/// doing so would lead to use-after-free and inconsistent states.
 pub struct Sha256State {
-    state: symcrypt_sys::SYMCRYPT_SHA256_STATE,
+    state: Pin<Box<Sha256InnerState>>,  
 }
 
+struct Sha256InnerState(symcrypt_sys::SYMCRYPT_SHA256_STATE);
+
+/// Creates a new instance of Sha256State, this must be called before other HashState functions can be called
 impl Sha256State {
     pub fn new() -> Self {
-        let mut instance = Sha256State {
-            state: symcrypt_sys::SYMCRYPT_SHA256_STATE::default(),
+        let mut instance =  Sha256State {
+            state: Box::pin(Sha256InnerState(symcrypt_sys::SYMCRYPT_SHA256_STATE::default())),
         };
-        unsafe {
-            symcrypt_sys::SymCryptSha256Init(&mut instance.state);
+        unsafe { // SAFETY: FFI calls
+            symcrypt_sys::SymCryptSha256Init(&mut instance.state.0);
         }
         instance
     }
 }
 
-impl Hash for Sha256State {
+impl HashState for Sha256State {
     type Result = [u8; SHA256_RESULT_SIZE];
 
     fn append(&mut self, data: &[u8]) {
-        unsafe {
+        unsafe { // SAFETY: FFI calls
             symcrypt_sys::SymCryptSha256Append(
-                &mut self.state,
+                &mut self.state.0,
                 data.as_ptr(),
                 data.len() as symcrypt_sys::SIZE_T,
             );
         }
     }
 
-    fn result(&mut self) -> Self::Result {
+    fn result(mut self) -> Self::Result {
         let mut result = [0u8; SHA256_RESULT_SIZE];
-        unsafe {
-            symcrypt_sys::SymCryptSha256Result(&mut self.state, result.as_mut_ptr());
+        unsafe { // SAFETY: FFI calls
+            symcrypt_sys::SymCryptSha256Result(&mut self.state.0, result.as_mut_ptr());
         }
         result
     }
+
+    fn copy(&self) -> Box<Self> {
+        let mut new_state = Sha256State {
+            state: Box::pin(Sha256InnerState(symcrypt_sys::SYMCRYPT_SHA256_STATE::default())),
+        };
+        unsafe { // SAFETY: FFI calls
+            symcrypt_sys::SymCryptSha256StateCopy(&self.state.0, &mut new_state.state.0);
+        }
+        Box::new(new_state)
+    }
 }
 
 impl Drop for Sha256State {
```

Comment:
>Sha256State [](http://example.com/codeflow?start=14&length=11)

Why the drop for Sha256State and not Sha256InnerState?

Reply:
I think that its cleaner, Sha256State is really just a wrapper for the Pin, it follows what we are doing in GCM as well

Reply:
Recommending getting the Drop closer to the inner object that could be included in other objects.

Rust ensures that drop is called for all inner objects.